File: Python/teil.py
from typing import List


def lade_teile_aus_tsv_file(tsv_file: str) -> List["Teil"]:
    teile = []
    with open(tsv_file, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            words = line.split()

            if line:  # leere Zeilen überspringen
                teil = Teil(
                    words[0], words[1], words[2], float(words[3]), float(words[4])
                )
                teile.append(teil)
    return teile


def baue_preis_dict(teile: List["Teil"]) -> dict:
    preise = {}
    for teil in teile:
        name = f"{teil.typ} {teil.bezeichnung}"
        preise[name] = teil.preis  # float
    return preise


class Teil:
    def __init__(
        self, typ, bezeichnung, einheit, gewicht, preis, komponenten=None
    ) -> None:
        self.typ = typ
        self.bezeichnung = bezeichnung
        self.einheit = einheit
        self.gewicht = gewicht
        self.preis = preis
        self.komponenten: List[Teil] = []

        if komponenten is None:
            komponenten = []
        self.komponenten = komponenten

    def __repr__(self):
        return f""" 
    Typ: {self.typ}
    Bez: {self.bezeichnung}
    Einheit: {self.einheit}
    Gewicht: {self.gewicht}
    Preis: {self.preis}
    Komponenten: {self.komponenten}
    """


def main():
    teile = lade_teile_aus_tsv_file("teil.dat")
    for teil in teile:
        print(teil)


if __name__ == "__main__":
    main()

File: C/teil.c
#include "teil.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

TeilListe lade_teile_aus_file(const char *filename) {
  TeilListe liste;
  liste.data = NULL;
  liste.count = 0;

  FILE *f = fopen(filename, "r");
  if (!f) {
    perror("Fehler beim Öffnen der Datei");
    return liste;
  }

  char line[256]; // gg
  while (fgets(line, sizeof(line), f)) {
    line[strcspn(line, "\r\n")] = 0; // Whitespace entfernen (wie strip())

    if (strlen(line) == 0) // leere Zeile überspringen
      continue;

    char *words[5]; // Wörter splitten
    int i = 0;
    char *token = strtok(line, " ");
    while (token && i < 5) {
      words[i++] = token;
      token = strtok(NULL, " ");
    }

    if (i < 5)
      continue; // fehlerhafte Zeile

    liste.data =
        realloc(liste.data,
                (liste.count + 1) * sizeof(Teil)); // Liste mit Teilen befuellen
    Teil *t = &liste.data[liste.count];
    snprintf(t->typ, sizeof(t->typ), "%s", words[0]);
    snprintf(t->bezeichnung, sizeof(t->bezeichnung), "%s", words[1]);
    snprintf(t->einheit, sizeof(t->einheit), "%s", words[2]);
    t->gewicht = atof(words[3]);
    t->preis = atof(words[4]);

    liste.count++;
  }

  fclose(f);
  return liste;
}

PreisDict baue_preis_dict(const TeilListe *liste) {
  PreisDict dict;
  dict.data = NULL;
  dict.count = 0;

  for (int i = 0; i < liste->count; i++) {
    const Teil *t = &liste->data[i];

    dict.data = realloc(dict.data, (dict.count + 1) * sizeof(PreisEintrag));
    PreisEintrag *e = &dict.data[dict.count];

    snprintf(e->name, sizeof(e->name), "%s %s", t->typ,
             t->bezeichnung); // name = "typ bezeichnung"
    e->preis = t->preis;

    dict.count++;
  }
  return dict;
}

double get_preis(const char *name, PreisDict *preise) {
  for (int i = 0; i < preise->count; i++) {
    if (strcmp(preise->data[i].name, name) == 0)
      return preise->data[i].preis;
  }
  return 0.0;
}

int main_teil(int argc, char *argv[]) {
  const char *filename = "teil.dat";
  TeilListe teile = lade_teile_aus_file(filename);

  for (int i = 0; i < teile.count; i++) {
    printf(
        "typ: %s\nbezeichnung: %s\neinheit: %s\ngewicht: %.2f\npreis: %.2f\n\n",
        teile.data[i].typ, teile.data[i].bezeichnung, teile.data[i].einheit,
        teile.data[i].gewicht, teile.data[i].preis);
  }

  free(teile.data);

  PreisDict p = baue_preis_dict(&teile);

  for (int i = 0; i < p.count; i++) {
    printf("%s -> %.2f\n", p.data[i].name, p.data[i].preis);
  }

  free(p.data);
  return EXIT_SUCCESS;
}
